nix/buildLeanPackage.nix

```diff
@@ -114,31 +146,39 @@ with builtins; let
     ${lean-emacs}/bin/emacs --eval "(progn (setq lean4-rootdir \"${lean}\") (require 'lean4-mode))" "$@"
   '';
   makeVSCodeWrapper = name: lean: writeShellScriptBin name ''
-    PATH=${lean}/bin:$PATH ${lean-vscode}/bin/code "$@"
+    PATH=${lean}/bin:$PATH LEAN_WORKER_PATH=${leanArg}/bin/lean ${lean-vscode}/bin/code "$@"
   '';
   printPaths = deps: writeShellScriptBin "print-paths" ''
     echo "${depRoot "print-paths" deps}"
-    echo ".:${lib.concatStringsSep ":" (map (dep: dep.src) (attrValues allDeps))}"
+    echo ".:${lib.concatStringsSep ":" (map (dep: dep.src) (attrValues allExternalDeps))}"
   '';
   makePrintPathsFor = deps: mods: printPaths deps // mapAttrs (_: mod: makePrintPathsFor (deps ++ [mod]) mods) mods;
-in rec {
-  inherit name lean deps allDeps print-lean-deps src;
   mods      = buildModAndDeps name {};
+  makeStaticLibDeriv = objects: runCommand "${name}-lib" { buildInputs = [ stdenv.cc.bintools.bintools ]; } ''
+    mkdir -p $out
+    ar Trcs $out/lib${name}.a ${lib.concatStringsSep " " (map (drv: "${drv}/${drv.oPath}") (attrValues objects))}
+  '';
+in rec {
+  inherit name lean deps staticLibDeps allExternalDeps print-lean-deps src mods;
   modRoot   = depRoot name [ mods.${name} ];
   cTree     = symlinkJoin { name = "${name}-cTree"; paths = map (mod: mod.c) (attrValues mods); };
   objects   = mapAttrs compileMod mods;
   oTree     = symlinkJoin { name = "${name}-oTree"; paths = (attrValues objects); };
-  staticLib = runCommand "${name}-lib" { buildInputs = [ stdenv.cc.bintools.bintools ]; } ''
-    mkdir $out
-    ar Trcs $out/lib${name}.a ${lib.concatStringsSep " " (map (drv: "${drv}/${drv.oPath}") (attrValues objects))}
+  staticLib = makeStaticLibDeriv objects;
+  plugin = runCommand "${name}.so" { buildInputs = [ stdenv.cc ]; } ''
+    mkdir -p $out/lib
+    ${leanc}/bin/leanc -fPIC -rdynamic -shared -x none \
+      -Wl,--whole-archive ${staticLib}/* -Wl,--no-whole-archive\
+      ${lib.concatStringsSep " " (map (d: "${d}/lib${d.name}.a") allPICStaticLibDeps)} \
+      -o $out/lib/${name}.so
   '';
   executable = runCommand executableName { buildInputs = [ stdenv.cc ]; } ''
     mkdir -p $out/bin
-    ${leanc}/bin/leanc -x none ${staticLib}/* -o $out/bin/${executableName}
+    ${leanc}/bin/leanc -x none ${staticLib}/* ${lib.concatStringsSep " " (map (d: "${d}/*.a") allStaticLibDeps)} -o $out/bin/${executableName}
   '';
 
   lean-package = writeShellScriptBin "lean" ''
-    LEAN_PATH=${modRoot}:$LEAN_PATH LEAN_SRC_PATH=${src}:$LEAN_SRC_PATH ${lean-final}/bin/lean "$@"
+    LEAN_PATH=${modRoot}:$LEAN_PATH LEAN_SRC_PATH=${src}:$LEAN_SRC_PATH ${lean-final}/bin/lean ${leanPluginFlags} "$@"
```

Comment:
As above, should probably not be inherited outside of the package

Reply:
Removed.